File: site_adapters/services/engine/scaffolds/snapshot.py
"""
Snapshot hook scripts for site-adapters (Python external mode).

Python snapshot scripts always run outside SingleFile. Use them for
pre-fetching, cookie/header/request URL handling, custom HTML generation, or
final file rewriting.

For DOM-level hooks that should run inside the SingleFile browser, use the
JavaScript scaffold `snapshot.js`. This Python scaffold is for external hooks
only.

Each script defines one or more hook functions. The function name is the hook
value from the adapter configuration. Define only the hooks you need.

Hook execution order:
    before scripts -> [replace script OR SingleFile engine] -> after scripts

--------------------------------------------------------------------------------
Config keys available in every hook
--------------------------------------------------------------------------------

  General:
    headers            dict[str,str]  HTTP request headers
    timeout            int|None       Timeout in seconds
    proxy              str|None       HTTP proxy URL
    request_url        str|None       Resolved request URL
    auth               dict           Merged auth config
    cookie             dict           Cookie configuration
    user_cookie        str|None       Best available cookie string

  Snapshot-specific:
    keep_elements      list[str]      CSS selectors to keep
    remove_elements    list[str]      CSS selectors to remove
    process_lazy_images bool|list[str]
    remove_classes     dict           CSS classes to remove
    set_styles         dict           Inline styles to set
    singlefile_args    dict           SingleFile CLI args
    toggles            dict           User-toggleable controls
    scripts            list[dict]     Script hooks configuration

--------------------------------------------------------------------------------
Project helpers commonly useful in custom Python hooks
--------------------------------------------------------------------------------

  # Browser engine selected by project settings (Chromium / CloakBrowser):
  # from site_adapters.services.engine.browser_provider import launch_browser, get_browser_config

  # Cookie conversion for Playwright contexts:
  # from site_adapters.services.auth.cookies import cookie_string_to_playwright_list

  # Stored credentials, with user-first/shared fallback helpers:
  # from site_adapters.services.auth.credentials import get_best_cookie, get_user_cookie, get_shared_cookie
  # from site_adapters.services.auth.credentials import get_best_header, get_best_token, get_best_basic_auth

  # OAuth2 access token cache/refresh:
  # from site_adapters.services.auth.oauth2 import get_valid_token

  # Resolve merged adapter configs for another URL/section:
  # from site_adapters.services.config.resolver import get_metadata_config, get_snapshot_config, get_reader_config

  # Delegate back to built-in engines from a replace hook:
  # from site_adapters.services.engine import create_snapshot, parse_metadata, run_script

--------------------------------------------------------------------------------
File handling
--------------------------------------------------------------------------------

  before: return an HTML string to feed to SingleFile, or None to let
          SingleFile fetch the URL normally.

  replace: write a complete HTML file to output_path.

  after: read and modify output_path in-place.

  The framework creates and cleans up temp files automatically.
"""


def before(url: str, config: dict) -> str | None:
    """
    Hook: before

    Executes before SingleFile. Return HTML to feed to SingleFile instead of
    re-fetching the URL; return None to let SingleFile fetch the URL normally.

    Example - expand collapsed content in the configured project browser:
        from site_adapters.services.engine.browser_provider import launch_browser

        browser = launch_browser(headless=True)
        playwright = getattr(browser, "__playwright__", None)
        try:
            page = browser.new_page()
            page.goto(config.get("request_url") or url, wait_until="networkidle")
            for btn in page.query_selector_all(".read-more-btn"):
                try:
                    btn.click()
                    page.wait_for_timeout(300)
                except Exception:
                    pass
            html = page.content()
        finally:
            browser.close()
            if playwright:
                playwright.stop()
        return html

    Example - return a fixed HTML document:
        return "<html><body><h1>Captured</h1></body></html>"
    """
    return None


def replace(url: str, config: dict, output_path: str) -> None:
    """
    Hook: replace

    Completely replaces the SingleFile engine. Write a complete,
    self-contained HTML file to `output_path`.

    Example - capture with the configured project browser:
        from site_adapters.services.engine.browser_provider import launch_browser

        browser = launch_browser(headless=True)
        playwright = getattr(browser, "__playwright__", None)
        try:
            page = browser.new_page()
            page.goto(config.get("request_url") or url, wait_until="networkidle")
            page.evaluate("window.scrollTo(0, document.body.scrollHeight)")
            page.wait_for_timeout(1000)
            html = page.content()
        finally:
            browser.close()
            if playwright:
                playwright.stop()

        with open(output_path, "w", encoding="utf-8") as f:
            f.write(html)

    Example - custom work then delegate to SingleFile:
        from site_adapters.services.engine import create_snapshot

        # ... do custom preprocessing ...
        create_snapshot(url, output_path, config)
    """
    pass


def after(output_path: str, config: dict) -> None:
    """
    Hook: after

    Executes after the snapshot HTML file is written. Modify the file
    at `output_path` in-place.

    Example - inject dark mode CSS:
        with open(output_path, "r+", encoding="utf-8") as f:
            html = f.read()
            html = html.replace(
                "</head>",
                "<style>body{background:#111;color:#eee}</style></head>",
            )
            f.seek(0)
            f.write(html)
            f.truncate()

    Example - rewrite CDN image URLs:
        import re

        with open(output_path, "r+", encoding="utf-8") as f:
            html = f.read()
            html = re.sub(
                r"https://cdn\\.example\\.com/",
                "https://example.com/",
                html,
            )
            f.seek(0)
            f.write(html)
            f.truncate()
    """
    pass

File: site_adapters/services/engine/scaffolds/snapshot.js
/**
 * Snapshot SingleFile browser hook (site-adapters).
 *
 * This template is the default for snapshot JavaScript hooks.
 * `before` runs inside the SingleFile browser context.
 * `after` runs against the saved snapshot HTML after SingleFile writes it.
 *
 * ---------------------------------------------------------------------------
 * Built-in engine declaration
 * ---------------------------------------------------------------------------
 *
 * const builtin_engine = "singlefile";
 *
 * Allowed values:
 *   "singlefile" - before runs inside SingleFile; after runs on saved HTML
 *   "" or null  - external Node mode (use snapshot_node.js)
 *   any other value is an error.
 *
 * This variable is only used by snapshot JavaScript scripts. Python snapshot
 * scripts are always external and do not declare it.
 *
 * ---------------------------------------------------------------------------
 * Runtime behavior
 * ---------------------------------------------------------------------------
 *
 * The framework wraps `before` into the SingleFile browser script. `after`
 * is executed separately against the saved HTML. The current implementation
 * uses Linkedom.
 *
 *   before(url, config) runs when `single-file-on-before-capture-request`
 *   fires, before SingleFile captures the page.
 *
 *   after(url, config) runs after the snapshot HTML file is written. DOM
 *   changes are serialized back into the saved file.
 *
 * `before` may be async; the framework calls `preventDefault()` and waits for
 * the matching `-response` event. `after` may also be async and is awaited by
 * the Node runner.
 *
 * `before` runs in the page, so browser DOM APIs are available. `after` runs
 * in Node with a Linkedom `document`/`window`, so DOM changes are possible but
 * Linkedom is not a real browser: property reflection is incomplete. Use
 * `setAttribute()`/`getAttribute()` for attributes that must persist, e.g.
 * `video.setAttribute("src", url)` rather than `video.src = url`. Node APIs
 * such as `require`, `fs`, and `process` are not exposed to user code. For
 * external Node hooks that need those APIs, use `snapshot_node.js`.
 *
 * ---------------------------------------------------------------------------
 * Config keys available in every hook
 * ---------------------------------------------------------------------------
 *
 * The framework injects the sanitized config and URL. Common keys:
 *
 *   headers            object         HTTP request headers
 *   timeout            number|null    Timeout in seconds
 *   proxy              string|null    HTTP proxy URL
 *   request_url        string|null    Resolved request URL
 *   user_cookie        string|null    Best available cookie string
 *   keep_elements      string[]       CSS selectors to keep
 *   remove_elements    string[]       CSS selectors to remove
 *   process_lazy_images boolean|string[]
 *   remove_classes     object         CSS classes to remove
 *   set_styles         object         Inline styles to set
 *   singlefile_args    object         SingleFile CLI args
 *   toggles            object         User-toggleable controls
 */

const builtin_engine = "singlefile";

async function before(url, config) {
  /**
   * Runs before SingleFile captures the page.
   * Modify the live DOM; changes are included in the snapshot.
   *
   * Example - expand collapsed content:
   *   document.querySelectorAll('.RichContent.is-collapsed').forEach((el) => {
   *     el.classList.remove('is-collapsed');
   *   });
   *
   * Example - lazy-load images from custom attributes:
   *   document.querySelectorAll('img[data-src]').forEach((img) => {
   *     if (!img.getAttribute('src')) {
   *       img.setAttribute('src', img.getAttribute('data-src'));
   *     }
   *   });
   */
}

async function after(url, config) {
  /**
   * Runs after the snapshot HTML file is written. DOM changes are saved back.
   * The current implementation runs in a Linkedom DOM, not a real browser:
   * property reflection is incomplete, so use setAttribute()/getAttribute()
   * for attributes that must persist.
   *
   * Example - inject dark mode:
   *   const style = document.createElement('style');
   *   style.textContent = 'body { background: #111; color: #eee; }';
   *   document.head.appendChild(style);
   */
}
